File: tests/fixtures/new-user-t03/validator.ts
// Executable fixture grammar; task adapters are implemented after contract freeze.
import { readFileSync } from "node:fs";
const contract = JSON.parse(readFileSync(new URL("./strict-contract.schema.json", import.meta.url), "utf8"));
type Rule = { ref?: string; anyOf?: Rule[]; enum?: unknown[]; type?: string; properties?: Record<string, Rule>; optional?: string[]; items?: Rule; min?: number; max?: number; unique?: boolean; id_prefixes?: string[]; format?: string };
const definitions = contract.definitions as Record<string, Rule>;
function require(condition: unknown): asserts condition { if (!condition) throw new Error("invalid_dto"); }
function check(rule: Rule, value: any): void {
  if (rule.ref) return check(definitions[rule.ref]!, value);
  if (rule.anyOf) {
    for (const option of rule.anyOf) { try { check(option, value); return; } catch {} }
    throw new Error("invalid_dto");
  }
  if (rule.enum) { require(rule.enum.some(item => item === value)); return; }
  switch (rule.type) {
    case "null": require(value === null); break;
    case "boolean": require(typeof value === "boolean"); break;
    case "integer": require(Number.isSafeInteger(value) && value >= rule.min! && value <= rule.max!); break;
    case "string": {
      require(typeof value === "string");
      const scalars = [...value];
      require(scalars.every(c => c.codePointAt(0)! !== 0 && !(c.codePointAt(0)! >= 0xd800 && c.codePointAt(0)! <= 0xdfff)));
      require(scalars.length >= (rule.min ?? 0) && scalars.length <= (rule.max ?? 1048576));
      if (rule.id_prefixes) require(new RegExp(`^(?:${rule.id_prefixes.join("|")})_[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$`).test(value));
      if (rule.format === "query") require(/[\p{L}\p{N}]/u.test(value) && new TextEncoder().encode(value).length <= 4096);
      if (rule.format === "timestamp") require(/^\d{4}-\d\d-\d\dT\d\d:\d\d:\d\dZ$/.test(value));
      break;
    }
    case "array":
      require(Array.isArray(value) && value.length >= rule.min! && value.length <= rule.max!);
      if (rule.unique) require(new Set(value.map(item => JSON.stringify(item))).size === value.length);
      value.forEach(item => check(rule.items!, item)); break;
    case "object":
      require(value !== null && typeof value === "object" && !Array.isArray(value));
      require(Object.keys(value).every(key => Object.prototype.hasOwnProperty.call(rule.properties!, key)));
      require(Object.keys(rule.properties!).every(key => rule.optional!.includes(key) || Object.prototype.hasOwnProperty.call(value, key)));
      Object.entries(value).forEach(([key, child]) => check(rule.properties![key]!, child));
      if ("complete" in value) require(value.complete === (value.next_cursor === null));
      if ("representative_capture_id" in value) require(value.representative_capture_id === value.capture_ids[0]);
      if ("record_type" in value) {
        if (value.record_type === "source") {
          require(value.record_id.startsWith("capture_") && value.record_id === value.revision_id && value.source_id !== null);
          require(value.provenance.capture_ids.length === 1 && value.provenance.capture_ids[0] === value.record_id);
        } else require(value.record_id.startsWith("page_") && value.revision_id.startsWith("revision_") && value.source_id === null);
      }
      if ("content" in value) {
        const length = new TextEncoder().encode(value.content.text).length;
        require(value.end_byte - value.start_byte === length && (value.complete || length > 0));
      }
      if ("required_grants" in value) require(value.required_grants.length === 1 && value.required_grants[0] === (contract.grants as Record<string, string>)[value.name]);
      if ("operations" in value) require(new Set(value.operations.map((item: any) => item.name)).size === value.operations.length);
      break;
    default: throw new Error("invalid_schema");
  }
}
export function validate(name: string, value: unknown): unknown { check(definitions[name]!, value); return value; }

File: tests/fixtures/new-user-t03/validator.py
"""Executable closed T03 DTO grammar. Contract proof, not a task implementation."""

import json
import re
from pathlib import Path
from typing import Any, NoReturn

CONTRACT = json.loads(Path(__file__).with_name("strict-contract.schema.json").read_text())


def parse(raw: str | bytes) -> Any:
    def pairs(items: list[tuple[str, Any]]) -> dict[str, Any]:
        result = {}
        for key, value in items:
            if key in result:
                raise ValueError("invalid_json")
            result[key] = value
        return result

    def reject(_: str) -> NoReturn:
        raise ValueError("invalid_json")

    decoded = raw.decode("utf-8") if isinstance(raw, bytes) else raw
    return json.loads(decoded, object_pairs_hook=pairs, parse_constant=reject)


def validate(name: str, value: Any) -> Any:
    def check(rule: dict[str, Any], value: Any) -> None:
        if "ref" in rule:
            return check(CONTRACT["definitions"][rule["ref"]], value)
        if "anyOf" in rule:
            for option in rule["anyOf"]:
                try:
                    check(option, value)
                    return
                except ValueError:
                    pass
            raise ValueError("invalid_dto")
        if "enum" in rule:
            if not any(type(value) is type(item) and value == item for item in rule["enum"]):
                raise ValueError("invalid_dto")
            return
        kind = rule["type"]
        types: dict[str, Any] = {
            "object": dict,
            "array": list,
            "string": str,
            "integer": int,
            "boolean": bool,
            "null": type(None),
        }
        if type(value) is not types[kind]:
            raise ValueError("invalid_dto")
        if kind == "object":
            props = rule["properties"]
            if set(value) - set(props) or set(props) - set(rule["optional"]) - set(value):
                raise ValueError("invalid_dto")
            for key, child in value.items():
                check(props[key], child)
            semantic(value)
        elif kind == "array":
            if not rule["min"] <= len(value) <= rule["max"]:
                raise ValueError("invalid_dto")
            if rule["unique"] and len({json.dumps(x, sort_keys=True) for x in value}) != len(value):
                raise ValueError("invalid_dto")
            for child in value:
                check(rule["items"], child)
        elif kind == "integer":
            if not rule["min"] <= value <= rule["max"]:
                raise ValueError("invalid_dto")
        elif kind == "string":
            if "\0" in value or any(0xD800 <= ord(c) <= 0xDFFF for c in value):
                raise ValueError("invalid_dto")
            if not rule.get("min", 0) <= len(value) <= rule.get("max", 1048576):
                raise ValueError("invalid_dto")
            if "id_prefixes" in rule:
                pattern = (
                    "(?:"
                    + "|".join(rule["id_prefixes"])
                    + ")_[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}"
                )
                if re.fullmatch(pattern, value) is None:
                    raise ValueError("invalid_dto")
            if rule.get("format") == "query" and (
                not any(c.isalnum() for c in value) or len(value.encode()) > 4096
            ):
                raise ValueError("invalid_dto")
            if (
                rule.get("format") == "timestamp"
                and re.fullmatch(r"\d{4}-\d\d-\d\dT\d\d:\d\d:\d\dZ", value) is None
            ):
                raise ValueError("invalid_dto")

    check(CONTRACT["definitions"][name], value)
    return value


def semantic(value: dict[str, Any]) -> None:
    def require(condition: bool) -> None:
        if not condition:
            raise ValueError("invalid_dto")

    if "complete" in value:
        require(value["complete"] == (value["next_cursor"] is None))
    if "representative_capture_id" in value:
        require(value["representative_capture_id"] == value["capture_ids"][0])
    if "record_type" in value:
        if value["record_type"] == "source":
            require(
                value["record_id"].startswith("capture_")
                and value["record_id"] == value["revision_id"]
                and value["source_id"] is not None
            )
            require(value["provenance"]["capture_ids"] == [value["record_id"]])
        else:
            require(
                value["record_id"].startswith("page_")
                and value["revision_id"].startswith("revision_")
                and value["source_id"] is None
            )
    if "content" in value:
        length = len(value["content"]["text"].encode())
        require(value["end_byte"] - value["start_byte"] == length)
        require(value["complete"] or length > 0)
    if "required_grants" in value:
        require(value["required_grants"] == [CONTRACT["grants"][value["name"]]])
    if "operations" in value:
        require(len({item["name"] for item in value["operations"]}) == len(value["operations"]))
